### ISSUE TITLE
An API for reading and writing to database is needed

### ISSUE THREAD

**Issue by jevers76** [dolson334/banterCustomer#27]
1.  I plan store our deposit information in the Notes field for reservations.   This will be a string of text formatted in a special way that will contain the following information.  I will make a program that will read the Notes and update the Notes for the following year
    a. The name of the deposit holder (this might be different or the same as the reservation holder)
    b. The amount of the deposit
    c. The date of the date slot (The dates change every year so when the deposit is transferred to the next year the dates will be adjusted for that year)
    d.  The cabin number
    e.  The date the deposit was first made
2.  We need a way to add gas charges to customer bills programmatically.  I plan to build a kiosk device that will be located by the gas pump for this.  I will need to be able to read cabin reservations, boat reservations and add a charge to a customer's bill.

**Comment by dolson334** [comment]
@jevers76 we can do this. Will you be using python to make the calls?

**Comment by jevers76** [comment]
Yes, python

**Comment by dolson334** [comment]
**Authentication Call** - returns a token that you can reuse for other calls. 

```
import requests
import json

url = "http://localhost:8080/login"

payload = json.dumps({
  "userName": "{yourUserName}",
  "password": "{yourPassword}"
})
headers = {
  'Content-Type': 'application/json'
}
response = requests.request("POST", url, headers=headers, data=payload)
```
the response will be like this, you then us the `token` as a bearer auth token for any other calls. 
```
[
    {
        "token": "eyJhbGciOiJIUzUxMiJ9.eyJqdGkiOiJ0ZXN0cmVzb3J0Iiwic3ViIjoiam9obkBzcHJpbmdiYXl0ZXN0LmNvbSIsImF1dGhvcml0aWVzIjpbIlJPTEVfUkVTT1JUX0FETUlOIl0sImlhdCI6MTY0NDcwMTE4OSwiZXhwIjoxNjQ0NzYxMTg5fQ.VOALjy94NAL6Ju2rlsDMU7_OL7MV5ajBIJ0ZoIC31DvPtlZ5209dp7RsF6cObsVfO9m9Tdm0X7ZOuLIduUAiqQ",
        "resortAlias": "testresort",
        "resortName": "Acorn Hill Resort"
    }
]
```

**Note Search API call** - searches for all notes that contain the passed in text. If no text is entered, all notes are returned. 

```
import requests

url = "http://localhost:8080/reservations/notes?resortAlias=testresort&searchText=search for"

payload = ""
headers = {
  'Authorization': 'Bearer {tokenFromAuthCall}'
}

response = requests.request("GET", url, headers=headers, data=payload)
```

the response will look like below. Notice we entered "search for" as the `searchText` parameter, which is in the `note` property below. 

```
[
    {
        "id": 1,
        "guestId": null,
        "reservationId": 18,
        "paymentId": null,
        "note": "search for a note",
        "endpoint": "notes",
        "objectType": "NOTE",
        "chargeType": null,
        "created": 1644701253800
    }
]
```

**Comment by jevers76** [comment]
I am not able to log on and get a token.  I originally received a 400 response using the code you sent, but I think the braces around the username and password need to be removed.  Now I get a 200 response which is OK.  Did you run the python code or is this just javascript converted to python?  I am using the kimps username and password and I know they work for logging into 'https://www.bantersoftware.com/admin'.  When I put "https://bantersoftware.com:8080/login" in my browser address bar I get an error message which might be expected I am not sure.

**Comment by dolson334** [comment]
@jevers76 these  calls worked for me. Do you have postman? It's a tool for calling APIs

**Comment by jevers76** [comment]
No I havn't heard of postman.  I will look it up.  How did you run the
python code?  from the cmd prompt or in your IDE?

**Comment by jevers76** [comment]
Dave,
It was working along.  The 200 response that printed was just the header of
the response.  The token is within the response.content part of the
response as response.content
John

**Comment by dolson334** [comment]
@jevers76 so you got the token?

**Comment by jevers76** [comment]
Yes, I got a token.  How long does it last?  Do I get a different one every time I post the username and password?  My next requests are not working yet.  I keep getting a 403 response

**Comment by jevers76** [comment]
Yes,  I did get the token.  I made additional comments within issue #27

**Comment by dolson334** [comment]
@jevers76 i will look at the logs shortly and see what the issue is. Your close I'm sure

**Comment by dolson334** [comment]
@jevers76 looking at the logs, it looks like you tried a couple times to update a note with the wrong id so the API will through a "no note found with the id".

**Comment by jevers76** [comment]
Yes, I was not replacing the 1 id with a valid id.  I am new to API calls and just learning. The id must be specified as part of the url prior to the ?.  When I changed the url to '   url = "https://bantersoftware.com:8080/reservations/18/notes/122?resortAlias=kimpscamp" it completes with a 200 response code, but it ends up deleting the note for  id 122!

What is the 18 in the string?
Why might it be deleting the not instead of replacing it with the one in the payload?
How will I link the id's to customers and reservations?

**Comment by dolson334** [comment]
@jevers76 the 18 is actually the `reservationId` (I should have stated that). That was just dummy data I was testing with. 

So basically, when you search for a note it will return the `reservationId` on the note object (like below). 
```
  {
        "id": 1,
        "guestId": null,
        "reservationId": 18,
        "paymentId": null,
        "note": "search for a note",
        "endpoint": "notes",
        "objectType": "NOTE",
        "chargeType": null,
        "created": 1644701253800
   }
```
You can then use this `reservationId` to query for the reservation, which will give you anything you might need pertaining to this note and reservation it is associated with (guest details, date booked, checkIn, checkOut, etc). 

```
import requests

url = "https://bantersoftware.com:8080/reservations/{reservationId}?resortAlias=kimpscamp"

payload = ""
headers = {
  'Authorization': 'Bearer {tokenFromAuthCall}'
}

response = requests.request("GET", url, headers=headers, data=payload)
```

**Comment by dolson334** [comment]
@jevers76 Ill add a logger so we can see. Ill let you know when its ready (roughly 5 minutes)

**Comment by jevers76** [comment]
OK I will be doing an update for id:126, reservationId:49.  Let me know when you are ready

**Comment by dolson334** [comment]
@jevers76 ok go for it.

**Comment by jevers76** [comment]
OK i sent the update, now the old note is gone

**Comment by jevers76** [comment]
Dave,  My mistake,  My search I was using was looking for the 'note' in the note and the update does not have this text so was not found making me think the note was deleted

**Comment by jevers76** [comment]
I havn't been able to get the add to work.  It shows a 415 error.
There are 19 reservations that show with with a custom search with a date range, yet there are 56 reservationId records.  Is this expected?  Is there some field that disables a reservationId record that I would use to filter out the ones that are still in use?
<img width="571" alt="image" src="https://user-images.githubusercontent.com/4566149/153777834-f444dd79-b7cb-4ab9-8ff2-7eafb6168254.png">

**Comment by dolson334** [comment]
@jevers76 can you send me the note add call you are making as well as the note body your sending?

on the reservations, yes, because the custom search filters out canceled reservation.

**Comment by jevers76** [comment]
I found the canceled field and now it finds the 18 reservations it should find. 

My add note  code is below.  It is the same as the update except for the letters 'add' instead of the id and the payload doesn't have an id key value.

def add(token, reservationId):
    import requests
    import json

    url = f"https://bantersoftware.com:8080/reservations/{reservationId}/notes/add?resortAlias=kimpscamp"
    payload = json.dumps({
      "note": "a new note is added here"
    })
    headers = {
      'Authorization': 'Bearer ' + token
    }
    response = requests.post(url, headers=headers, data=payload)
    return response

**Comment by jevers76** [comment]
It needed the Content-Type in the header.  Now it is working.  Thanks.
